### PR TITLE
feat: add support for nextcloud 30

### PR DESCRIPTION
**Pull request by EsadCetiner**
This PR adds support for Nextcloud 30, adds tests for some existing rules and refactors some existing rules.

### CODE REVIEW

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 1]
```diff
@@ -1057,38 +1075,28 @@ SecRule REQUEST_FILENAME "@endsWith /index.php" \
     nolog,\
     ver:'nextcloud-rule-exclusions-plugin/1.2.0',\
     chain"
-    SecRule ARGS:dbtype "@rx ^(?:my|pg)sql$" \
+    SecRule ARGS:dbtype "@rx ^(?:my|pg|sq)sql$" \
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:1078]: Is `sqsql` really a valid option? Shouldn't it be `sqlite`?

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 2]
```diff
@@ -1913,6 +1904,26 @@ SecRule REQUEST_FILENAME "@rx /apps/deck/stacks/[0-9]+/reorder$" \
     ver:'nextcloud-rule-exclusions-plugin/1.2.0',\
     setvar:'tx.allowed_methods=%{tx.allowed_methods} PUT'"
 
+# Editing/deleting a comment on a card
+SecRule REQUEST_FILENAME "@rx /ocs/v[0-9\.]+\.php/apps/deck/api/v[0-9\.]+/cards/[0-9]+/comments/[0-9]+$" \
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:1908]: `.` is treated as a literal in character classes and doesn't need to be escaped. There are a couple more of these, but I will only suggest it here.
```suggestion
SecRule REQUEST_FILENAME "@rx /ocs/v[0-9.]+\.php/apps/deck/api/v[0-9.]+/cards/[0-9]+/comments/[0-9]+$" \
```

File: plugins/nextcloud-rule-exclusions-before.conf [hunk 3]
```diff
@@ -1998,7 +2009,8 @@ SecRule REQUEST_FILENAME "@rx /ocs/v[0-9\.]+\.php/references/resolve$" \
     ctl:ruleRemoveTargetById=932200;ARGS:reference,\
     ctl:ruleRemoveTargetById=942430;ARGS:reference,\
     ctl:ruleRemoveTargetById=942431;ARGS:reference,\
-    ctl:ruleRemoveTargetById=942432;ARGS:reference"
+    ctl:ruleRemoveTargetById=942432;ARGS:reference,\
+    ctl:ruleRemoveTargetById=921151;ARGS_GET:reference"
```
Comment (theseion) [inline plugins/nextcloud-rule-exclusions-before.conf:2013]: Why `ARGS_GET` here?

File: tests/regression/nextcloud-rule-exclusions-plugin/9508170.yaml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                              <nc:trashbin-deleted-by-display-name />
+                              <d:getcontentlength /> <d:getcontenttype /> <d:getetag /> <d:getlastmodified /> <d:creationdate /> <d:displayname /> <d:quota-available-bytes />
+                               <d:resourcetype /> <nc:has-preview /> <nc:is-encrypted /> <nc:mount-type /> <oc:comments-unread /> <oc:favorite /> <oc:fileid /> <oc:owner-display-name /> <oc:owner-id />
+                               <oc:permissions /> <oc:size /> <nc:hidden /> <nc:is-mount-root /> <nc:metadata-files-live-photo /> <nc:reminder-due-date /> <nc:note /> <nc:sharees />
+                               <nc:share-attributes /> <oc:share-types /> <ocs:share-permissions />
+                       </d:prop>
+               </d:propfind>
+            version: HTTP/1.1
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508170.yaml:39]: Careful with block scalars. The trailing newline would change your regular expression to `...|id "921110"\n`.
```suggestion
            no_log_contains: |-
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml [hunk 5]
```diff
@@ -67,11 +68,50 @@ tests:
...
+            method: POST
+            uri: /index.php/login/confirm
+            data: |
+              {"password": "<script>"}
             version: HTTP/1.1
           output:
-            no_log_contains: |-
-              id "920230"|id "932190"|id "932200"|id "942431"|id "942432"
+            no_log_contains: id "941101"
+  - test_title: 9508400-6
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml:99]: This test looks like a duplicate of the test above.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508984.yaml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            headers:
+              Host: localhost
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: DELETE
+            uri: /apps/mail/api/messages/138/tags/%24test
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"`
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508984.yaml:39]:
```suggestion
            no_log_contains: id "911100"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 7]
```diff
@@ -0,0 +1,137 @@
+---
+meta:
+  author: "Esad Cetiner"
+  description: |
+    Nextcloud Mail
+    Note this test also applies for rule 9508988 (Which allows com and inc TLDs)
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:6]:
```suggestion
    Note: this test also applies for rule 9508988 (Which allows com and inc TLDs)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            headers:
+              Host: localhost
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: DELETE
+            uri: /apps/mail/api/internalAddress/test%40example.com?type=individual
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"`
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:41]:
```suggestion
            no_log_contains: id "911100"
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/trustedsenders/example.com?type=domain
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"
+  - test_title: 9508987-6
+    desc: Adding a domain from the trusted senders list
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:91]:
```suggestion
    desc: Removing a domain from the trusted senders list (.com)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: DELETE
+            uri: /apps/mail/api/trustedsenders/example.com?type=domain
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"
+  - test_title: 9508987-7
+    desc: Adding a domain to the trusted senders list
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:107]:
```suggestion
    desc: Adding a domain to the trusted senders list (.inc)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: DELETE
+            uri: /apps/mail/api/internalAddress/test%40example.inc?type=individual
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"
+  - test_title: 9508987-5
+    desc: Adding a domain to the trusted senders list
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:75]:
```suggestion
    desc: Adding a domain to the trusted senders list (.com)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/internalAddress/test%40example.com?type=individual
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"
+  - test_title: 9508987-2
+    desc: Removing internal addresses
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:27]:
```suggestion
    desc: Removing internal addresses (.com)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+meta:
+  author: "Esad Cetiner"
+  description: |
+    Nextcloud Mail
+    Note this test also applies for rule 9508988 (Which allows com and inc TLDs)
+  enabled: true
+  name: 9508987.yaml
+tests:
+  - test_title: 9508987-1
+    desc: Adding internal addresses
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:11]:
```suggestion
    desc: Adding internal addresses (.com)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: DELETE
+            uri: /apps/mail/api/internalAddress/test%40example.com?type=individual
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"`
+  - test_title: 9508987-3
+    desc: Adding internal addresses
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:43]:
```suggestion
    desc: Adding internal addresses (.inc)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/internalAddress/test%40example.inc?type=individual
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"
+  - test_title: 9508987-4
+    desc: Removing internal addresses
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:59]:
```suggestion
    desc: Removing internal addresses (.inc)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+              User-Agent: OWASP CRS test agent
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/trustedsenders/example.inc?type=domain
+            version: HTTP/1.1
+          output:
+            no_log_contains: id "911100"
+  - test_title: 9508987-8
+    desc: Removing a domain from the trusted senders list
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508987.yaml:123]:
```suggestion
    desc: Removing a domain from the trusted senders list (.inc)
```

File: tests/regression/nextcloud-rule-exclusions-plugin/9508989.yaml [hunk 17]
```diff
@@ -25,3 +25,23 @@ tests:
...
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+              Content-Type: application/json
+            port: 80
+            method: PUT
+            uri: /apps/mail/api/accounts/2/aliases/2/signature
+            data: |
+             {"signature":"<p style=\"margin:0;\">t</p>"}
+            version: HTTP/1.1
+          output:
+            no_log_contains: |
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508989.yaml:46]:
```suggestion
            no_log_contains: |-
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 1 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:1078]: Good catch, I'll add a test for that

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 3 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf:2013]: The rule checks for `ARGS_GET` and not `ARGS` https://github.com/coreruleset/coreruleset/blob/d33dad89052ecc531641b047efcaae339c254362/rules/REQUEST-921-PROTOCOL-ATTACK.conf#L321

File: tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml (hunk 5 above)
Comment (EsadCetiner) [reply tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml:99]: I'll remove the tests that check without pretty URLs, they are mostly redundant (I think pretty URLs are the default now in Nextcloud anyways). These 2 tests were meant to check both with and without pretty URLs

File: plugins/nextcloud-rule-exclusions-before.conf (hunk 1 above)
Comment (EsadCetiner) [reply plugins/nextcloud-rule-exclusions-before.conf]: Fixed. I also found another false positive when working on this.

File: tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml [hunk 18]
```diff
@@ -6,7 +6,7 @@ meta:
   name: 9508400.yaml
 tests:
   - test_title: 9508400-1
-    desc: Disable OWASP CRS for password with pretty URLs
+    desc: Logging into Nextcloud
```
Comment (theseion) [inline tests/regression/nextcloud-rule-exclusions-plugin/9508400.yaml:9]:
```suggestion
    desc: Logging in to Nextcloud
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Review by theseion — APPROVED** [review]

**Pull request merged**